test/src/vfs_test_helper.cc

```diff
@@ -0,0 +1,282 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines some vfs-specific test suite helper functions.
+ */
+#include "catch.hpp"
+
+#include "test/src/helpers.h"
+#include "test/src/vfs_test_helper.h"
+
+#ifdef _WIN32
+#include "tiledb/sm/filesystem/win.h"
+#else
+#include "tiledb/sm/filesystem/posix.h"
+#endif
+
+namespace tiledb {
+namespace test {
+
+Status SupportedFsS3::prepare_config(
+    tiledb_config_t* config, tiledb_error_t* error) {
+#ifndef TILEDB_TESTS_AWS_S3_CONFIG
+  REQUIRE(
+      tiledb_config_set(
+          config, "vfs.s3.endpoint_override", "localhost:9999", &error) ==
+      TILEDB_OK);
+  REQUIRE(
+      tiledb_config_set(config, "vfs.s3.scheme", "https", &error) == TILEDB_OK);
+  REQUIRE(
+      tiledb_config_set(
+          config, "vfs.s3.use_virtual_addressing", "false", &error) ==
+      TILEDB_OK);
+  REQUIRE(
+      tiledb_config_set(config, "vfs.s3.verify_ssl", "false", &error) ==
+      TILEDB_OK);
+  REQUIRE(error == nullptr);
+#endif
+
+  return Status::Ok();
+}
+
+Status SupportedFsS3::init(tiledb_ctx_t* ctx, tiledb_vfs_t* vfs) {
+  int is_bucket = 0;
+  int rc = tiledb_vfs_is_bucket(ctx, vfs, S3_BUCKET.c_str(), &is_bucket);
+  REQUIRE(rc == TILEDB_OK);
+  if (!is_bucket) {
+    rc = tiledb_vfs_create_bucket(ctx, vfs, S3_BUCKET.c_str());
+    REQUIRE(rc == TILEDB_OK);
+  }
+  return Status::Ok();
+}
+
+Status SupportedFsS3::close(tiledb_ctx_t* ctx, tiledb_vfs_t* vfs) {
+  int is_bucket = 0;
+  int rc = tiledb_vfs_is_bucket(ctx, vfs, S3_BUCKET.c_str(), &is_bucket);
+  CHECK(rc == TILEDB_OK);
+  if (is_bucket) {
+    CHECK(tiledb_vfs_remove_bucket(ctx, vfs, S3_BUCKET.c_str()) == TILEDB_OK);
+  }
+  return Status::Ok();
+}
+
+std::string SupportedFsS3::temp_dir() {
+  const std::string S3_PREFIX = "s3://";
+  const std::string S3_BUCKET = S3_PREFIX + random_name("tiledb") + "/";
+  const std::string S3_TEMP_DIR = S3_BUCKET + "tiledb_test/";
+  std::string vfs_helper_temp_dir = S3_TEMP_DIR;
```

Comment:
We have all of these as member variables, we should not be redeclaring them as locals. See my comment on initializing these in the constructor. This routine should just be:
```
std::string SupportedFsS3::temp_dir() {
  return vfs_helper_temp_dir_;
}
```